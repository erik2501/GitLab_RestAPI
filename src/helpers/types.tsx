import { NamedExoticComponent } from "react";

export type Commit = {
    id: string;
    short_id: string;
    created_at: string;
    parent_ids: string[];
    title: string;
    message: string;
    author_name: string;
    author_email: string;
    authored_date: string;
    committer_name: string;
    committer_email: string;
    committed_date: string;
    trailers: unknown;
    web_url: string;
}

type Author = {
    id: number,
    username: string,
    name: string,
    state: string,
    avatar_url: string,
    web_url: string
}

export type User = {
    id: number;
    username: string;
    name: string;
    state: string;
}


export type Issue = {
    id: number,
    iid: number,
    project_id: number,
    title: string,
    description: string, 
    state: string,
    created_at: string,
    updated_at: string,
    closed_at: string, 
    closed_by: User,
    labels: string[], 
    milestone: string, 
    assignees: string[], 
    author: Author,
    type: string,
    assignee: string, 
    user_notes_count: number
}

type Namespace = {
    name: string
}

export type Project = {
    id: number,
    description: string,
    name: string,
    avatar_url: string //"https://gitlab.stud.idi.ntnu.no/uploads/-/system/project/avatar/17430/pingvin.png"
    namespace: Namespace
}

export type LoginInfo = {
    projectID: string | null,
    token: string | null
}