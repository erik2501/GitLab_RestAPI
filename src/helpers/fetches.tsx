import { Commit } from "./types";


export async function getCommits() {
    const response = await fetch("https://gitlab.stud.idi.ntnu.no/api/v4/projects/17430/repository/commits", {
        headers:
        {
            Authorization: "Bearer glpat-n3y-kCt83mAmv5KK63js"
        }
    })
    if (response.ok) {
        return await response.json() as Commit[];
    } else {
        console.log('something went wrong')
    }
}


