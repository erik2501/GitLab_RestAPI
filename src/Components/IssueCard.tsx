import { Issue } from '../helpers/types';

function IssueCard({ issue }: { issue: Issue }) {


    return (
        <div className='commitCard'>
            <h5>state: {issue.state}</h5>
            <h5>{
                issue.closed_by ?
                    "closed at: " + (new Date(issue.closed_at)).toLocaleDateString('en-GB') :
                    "created at: " + (new Date(issue.created_at)).toLocaleDateString('en-GB')}</h5>
            <h2>{issue.title}</h2>
            <h3>{"Created by: " + issue.author.name} </h3>
            <h3>{issue.closed_by ? "Closed by: " + issue.closed_by.name : "Closed by: not yet closed"} </h3>
        </div>
    );


};

export default IssueCard;