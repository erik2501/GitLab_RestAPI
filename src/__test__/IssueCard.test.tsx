import IssueCard from '../Components/IssueCard';
import { Issue } from '../helpers/types';
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react';

const dummyAuthor = {
    id: 0,
    username: '',
    name: '',
    state: '',
    avatar_url: '',
    web_url: ''
}

const dummyUser = {
    id: 0,
    username: '',
    name: 'userName',
    state: ''
}



const dummyIssue: Issue = {
    id: 0,
    iid: 0,
    project_id: 0,
    title: 'title',
    description: 'description',
    state: 'closed',
    created_at: "2021-05-02T18:02:30.000-05:00",
    updated_at: "2021-05-02T18:02:30.000-05:00",
    closed_at: "2021-05-02T18:02:30.000-05:00",
    closed_by: dummyUser,
    labels: [''],
    milestone: '',
    assignees: [''],
    author: dummyAuthor,
    type: '',
    assignee: '',
    user_notes_count: 0
}


it('renders when passed an issue', () => {
    const tree = renderer.create((<IssueCard issue={dummyIssue} />)).toJSON();
    expect(tree).toMatchSnapshot();
});


test('should return date in nice format', () => {
    const { container } = render(<IssueCard issue={dummyIssue} />);
    screen.getByText('closed at: 03/05/2021')
});
