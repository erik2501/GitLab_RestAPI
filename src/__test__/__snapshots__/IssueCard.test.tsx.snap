// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when passed an issue 1`] = `
<div
  className="commitCard"
>
  <h5>
    state: 
    closed
  </h5>
  <h5>
    closed at: 03/05/2021
  </h5>
  <h2>
    title
  </h2>
  <h3>
    Created by: 
     
  </h3>
  <h3>
    Closed by: userName
     
  </h3>
</div>
`;
